kitchen/src/kitchen/core.clj

```diff
@@ -413,68 +407,83 @@
-                                       (reduce str))]
-                  (async/>! out string-data))
-
-                (and (string? in-data) (str/starts-with? in-data "bytes-") (> (count in-data) (count "bytes-")))
-                (let [num-bytes-str (subs in-data (count "bytes-"))
-                      num-bytes-int (Integer/parseInt num-bytes-str)
-                      byte-data (byte-array (take num-bytes-int (cycle (range 103))))]
-                  (async/>! out byte-data))
-
-                :else
-                (async/>! out in-data))
-              (recur))))))))
-
-(defn basic-auth-middleware [username password handler]
-  (fn [{:keys [uri] :as req}]
+  (printlog request "Received websocket request:" request)
+  (swap! pending-ws-requests inc)
+  (async/go
+    (async/>! out "Connected to kitchen")
+    (loop []
+      (let [in-data (async/<! in)]
+        (printlog request "Received data on websocket:" in-data)
+        (if (or (str/blank? (str in-data)) (= "exit" in-data))
+          (do
+            (async/>! out "bye")
+            (printlog request "Closing connection.")
+            (async/close! out)
+            (swap! pending-ws-requests dec))
+          (do
+            (cond
+              (instance? ByteBuffer in-data)
+              (let [response-bytes (byte-array (.remaining in-data))]
+                (.get in-data response-bytes)
+                (async/>! out response-bytes))
+
+              (= "request-info" in-data)
+              (async/>! out (-> request request-info-handler :body))
+
+              (= "kitchen-state" in-data)
+              (async/>! out (-> request state-handler :body))
+
+              (and (string? in-data) (str/starts-with? in-data "chars-") (> (count in-data) (count "chars-")))
+              (let [num-chars-str (subs in-data (count "chars-"))
+                    num-chars-int (Integer/parseInt num-chars-str)
+                    chars (map char (range 65 91))
+                    string-data (->> (repeatedly #(rand-nth chars))
+                                     (take num-chars-int)
+                                     (reduce str))]
+                (async/>! out string-data))
+
+              (and (string? in-data) (str/starts-with? in-data "bytes-") (> (count in-data) (count "bytes-")))
+              (let [num-bytes-str (subs in-data (count "bytes-"))
+                    num-bytes-int (Integer/parseInt num-bytes-str)
+                    byte-data (byte-array (take num-bytes-int (cycle (range 103))))]
+                (async/>! out byte-data))
+
+              :else
+              (async/>! out in-data))
+            (recur)))))))
+
+(defn basic-auth-middleware
+  "Adds support for basic authentication when both the provided username and password are not nil.
+  /status urls always bypass basic auth check."
+  [username password handler]
+  (when (not (and username password))
+    (log/info "basic authentication is disabled since username or password is missing"))
```

Comment:
I know this behavior wasn't introduced in this change but:
Why do we disable auth if username and password are missing?  Doesn't that mean if we ever stop passing auth info to kitchen it will be silently ignored?